llvm/lib/YkIR/YkIRWriter.cpp

```diff
@@ -1640,23 +1640,71 @@ class YkIRWriter {
     OutStreamer.emitInt8(Flags);
   }
 
+  // Check if the user did anything invalid in a yk_outline function.
```

Comment:
I don't think this is recursing into the call graph? Indeed, what happens if you have an indirect function call which then calls `promote`?

Reply:
Just so that we're definitely on the same page: what you're saying is that you can avoid a global analysis? And that this sort of thing is fine because even though there's an intermediate outlinable function, the `promote` is in the not-outlinable (do we have a name for that concept?!) function:

```rust
attr(outline)
fn f() { g(); }

g() { yk_promote(...) }
```

In other words, we will track the `yk_promote` in `g()` even when `g` is called from `f`?

Reply:
Correct. As long as `g()` isn't inlined into `f()`, but we will detect that and error out (as the added tests show).

This is just all about preventing "hidden" `yk_promote()`  (and friends) in functions that the trace builder will need to skip over, but which has no IR for.

Reply:
Got it.

[MInor note for the future to help a dummy like me. What I'd say is this commit enables a static, local analysis that outline functions can't directly call `yk_promote`. The -- at first to me surprising, but now I buy it! -- implication is that we need do neither a global nor a dynamic analysis because non-outline functions -- no matter where they come in the dynamic call graph -- can safely call `yk_promote` and we'l handle them properly.]